machine.go

```diff
@@ -501,11 +515,26 @@ func (m *Machine) startup(command string, extracontent [][2]string) (int, error)
 		"-enable-kvm",
 		"-kernel", "/boot/vmlinuz-" + kernelRelease,
 		"-initrd", InitrdPath,
-		"-nographic",
 		"-no-reboot"}
 	kernelargs := []string{"console=ttyS0", "quiet", "panic=-1",
 		"systemd.unit=fakemachine.service"}
 
+	if m.showBoot {
+		qemuargs = append(qemuargs, "-nographic")
+	} else {
+		qemuargs = append(qemuargs,
+			"-display", "none",
+			"-device", "virtio-serial",
+			"-chardev", "null,id=for-ttyS0",
+			"-chardev", "null,id=for-ttyS1",
+			"-chardev", "null,id=for-hvc0",
+			"-chardev", "stdio,id=for-hvc1",
+			"-device", "virtconsole,chardev=for-hvc0",
+			"-device", "virtconsole,chardev=for-hvc1",
+			"-serial", "chardev:for-ttyS0",
+			"-serial", "chardev:for-ttyS1")
```

Comment:
Would be nice to have some explanation of how this works as it's not entirely obvious to me (either in code comment or commit message).  I guess you're redirecting ttyS0/ttyS1 to /dev/null, and adding hvc consoles. But why ttyS1 as well, and similar why two hvc consoles rather then just one?

Reply:
The qemu command-line is basically a disaster area, with several layers of backwards compatibility where old special-case designs have become aliases for something newer and more general. `-nographic` is a shortcut for something resembling `-display none -chardev stdio,mux=on,id=foo -serial chardev:foo -mon chardev:foo` (possibly over-simplifying, I think `-nographic` might actually include a parallel port in the multiplexed character device too).

`-chardev null` isn't *exactly* `/dev/null`, but it's the equivalent in qemu code.

ttyS1 is unnecessary, I can drop that. I briefly thought I was going to have to use ttyS1, like autopkgtest-virt-qemu does, but then I got hvc1 working.

systemd special-cases the first virtualized console to put a getty on it automatically:

> "systemd will also spawn a login prompt on the first special VM console (that's /dev/hvc0, /dev/xvc0, /dev/hvsi0), if the system is run in a VM that provides these devices" —http://0pointer.de/blog/projects/serial-console.html

which is why I avoided using it; perhaps fakemachine is already blacklisting the relevant generator, but it seems best to avoid hvc0 for purposes that aren't actually the VM's interactive console.

Reply:
One thing we could maybe consider would be to use `console=hvc0`, so that there are three separate streams that you can have logged (or splatted onto stderr, as appropriate):

* ttyS0: BIOS messages corrupting your terminal, bootloader if used (but you're using direct kernel boot so that isn't applicable), and possibly the really early stages of Linux boot
* hvc0: systemd boot messages (I think these avoid corrupting the terminal state, although I'd need to test)
* hvc1: fakemachine/debos payload, aka what you actually wanted to see

Reply:
For now I've changed this to:

* ttyS0: BIOS, bootloader if we used one, Linux, initramfs (?), systemd
* hvc0: reserved but unused (systemd might put a getty here, but if it does, we ignore it)
* hvc1: payload

Reply:
To prevent systemd to put a getty there you can mask serial-getty@hvc0.service that's what we do for ttyS0

Reply:
To confirm, you would prefer there to be only two separable streams: all the console stuff (from BIOS up to systemd), and the payload?